### ISSUE TITLE
Fails hard when provisioning new user

### ISSUE THREAD

**Issue by thijskh** [OpenConext/OpenConext-engineblock#298]
This is with EB 5.0 ß 4 but can also be reproduced with 5.0 ß 3.

Logging in to EB as a user that has no row yet in `users` fails with a hard error:
```
Catchable fatal error: Object of class OpenConext\Component\EngineBlockMetadata\Service could not be converted to string in /opt/openconext/OpenConext-engineblock-5.0.0-beta4/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php on line 70
```

Interestingly enough it seems that EB tries to interpolate something into this query when it happens: `UPDATE sso_provider_roles SET certificates = ?, name_id_formats = ?, response_processing_service_binding = ?, allowed_idp_entity_ids = ?, requested_attributes = ? WHERE id = ?`. EB should probably never update sso_provider_roles during the login flow in the first place.

This might be related to the fact that `class UserRepository extends EntityRepository` as introduced in 1cd9adac19885137579b13bd2074efec7eca1bae?

**Comment by DRvanR** [comment]
Unfortunately I seem not to be able to reproduce this locally... However, some investigation does seem to indicate a possible cause:

It seems that this was existing, but unintended and untriggered behavior. What seems to be the case is that some code actually alters the `requestedAttributes` and other properties as listed on the ServiceProvider object (which is the case in some places, e.g. [here][2]) and that that causes Doctrine to detect and attempt to persist these changes when creating the new user.

A bit more detail: 
When a new user is registered, the user is persisted and the changes are committed. This causes the UnitOfWork to calculate all changes that occurred for registered Entities (all entities are tracked in an IdentityMap). If an Entity (such as a ServiceProvider) is loaded and altered before the new User is created and persisted, this means that the UnitOfWork would detect this Entity as "dirty", which means it differs from the Database version. Once the commit is made (the `flush` call), all changes are persisted to the database. Previously, there was no flush call made during a login, only when updating the metadata (metadata push). This means that any changes made to an Entity would be lost. This is now no longer the case, as the changes are committed (or at least, attempted to) when a new user is registered. Therefor it seems that this behavior existed, but was simply not triggered due to not committing changes to the database.

It seems the commit that is referenced indeed causes what is happening here, but only as an accidental side-effect. This also explains why this was not apparent in the dev envs: apparently the configuration there is such that there are no changes made to the ServiceProvider Entity, thus not triggering this effect.

In order to resolve this issue, all changes made to an Entity that are not intended to be persisted should no longer be made to that Entity but rather to a different object derived from that Entity that allows safe updates of data without polluting the Entity. An alternative solution that might be possible (have to consider metadata push with this) is to mark the ServiceProvider and IdentityProvider Entities as [readonly][3]. The latter could be a quick fix, the former would be the proper fix. 

As an FYI: The fact that the `UserRepository` extends the (Doctrine) `EntityRepository` is [required to be able to create custom repositories][1] - it does not influence Service Providers or Identity Providers entities as those are governed by a different repository (yet are tracked in the same IdentityMap).

[1]: http://doctrine-orm.readthedocs.io/projects/doctrine-orm/en/latest/reference/working-with-objects.html#custom-repositories
[2]: https://github.com/OpenConext/OpenConext-engineblock/blob/f08e3beb2a563f671129a9f540be65bc38e20fe3/library/EngineBlock/Corto/Module/Service/Metadata/ArpRequestedAttributes.php#L27
[3]: http://doctrine-orm.readthedocs.io/projects/doctrine-orm/en/latest/reference/improving-performance.html#read-only-entities

**Comment by thijskh** [comment]
Thanks for the thorough analysis. Of course I prefer the "proper" fix, but I'm not sure how much time that would take. Any idea on that? We want to move forward with 5.0 soon, so I'm guessing that we might prefer the quickfix now and postpone the properfix to 5.1. What do you think?

**Comment by DRvanR** [comment]
My only concern with the quickfix would be the metadata push. That uses some specific functionality to overwrite already configured services with the new configuration, and I'm not sure if the quickfix would allow that to still work - this will have to be tested. I will try to make a patch with the quickfix this week.

An alternative solution (should the quickfix not work) could also be to change the way changes are detected to [require explicit persist calls for services][1]. Codewise this would also be simple to achieve.

I do agree that the proper fix would be too much to implement right now.

Will keep you posted as to when I've got a patch ready for testing.

[1]: http://doctrine-orm.readthedocs.io/projects/doctrine-orm/en/latest/reference/change-tracking-policies.html#deferred-explicit

**Comment by thijskh** [comment]
Alright, thanks!

**Comment by quartje** [comment]
Adding a new sso_provider_roles_eb5 table in the database has fixed this issue.

**Comment by DRvanR** [comment]
@thijskh @quartje: What is the course of action to follow? Merge https://github.com/OpenConext/OpenConext-engineblock-metadata/pull/9, release the lib, require released version, and introduce an additional BC break (the requirement of a new metadata table)? Or should we add a patch to EB5 as well that creates said table and populates it with the data from the current table?

**Comment by thijskh** [comment]
We've discussed this briefly and prefer the latter approach. It's more clean so the right way to go for other users of EB.

**Comment by DRvanR** [comment]
Resolved through #301 and later #306
